File: src/Chap9StudentLists.java
import java.security.InvalidParameterException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

/**
 * A class to keep track of lists of students. Instances get an implicit ID via their insertion into the
 * static ArrayList "classes," with which a new class can be created
 */
public class Chap9StudentLists {
    static ArrayList<Chap9StudentLists> classes = new ArrayList<>(8);
    private final HashSet<String> students;

    /**
     * @param students
     */
    public Chap9StudentLists(HashSet<String> students) {
        this.students = students;
        classes.add(this);
    }

    /**
     * Constructor from existing class
     *
     * @param id               to pull starting class list from
     * @param studentsToAdd    set of students to add to existing class list
     * @param studentsToRemove set of students to remove from existing class list
     * @throws InvalidParameterException if class at index id is not found
     */
    public Chap9StudentLists(Integer id, Set<String> studentsToAdd, Set<String> studentsToRemove) throws InvalidParameterException {
        try {
            students = new HashSet<>(classes.get(id).students);
        } catch (IndexOutOfBoundsException e) {
            throw new InvalidParameterException("No class with id " + id + " exists.");
        }

        students.removeAll(studentsToRemove);
        students.addAll(studentsToAdd);

        classes.add(this);
    }

    /**
     * Getter
     *
     * @return students
     */
    public HashSet<String> getStudents() {
        return students;
    }

}

File: src/Chap9StudentLists.py
from typing import Iterable, Set


# A class to keep track of lists of students. Instances get an implicit ID via their insertion into the
# list of classes, with which a new class can be created
class Chap9StudentList:
    classes = []

    # Create new instance from a Set of students
    def __init__(self, students: Set):
        self.students = students
        self.classes.append(self)

    # Create new instance from existing instance, if available
    @classmethod
    def from_existing(cls, id, students_to_remove: Iterable, students_to_add: Iterable):
        try:
            existing_class = cls.classes[id]
        except IndexError:
            raise Exception(f"No class with id {id} found")

        students = existing_class.students.copy() - students_to_remove + students_to_add

        cls(students)  # calls __init__()
